zip-0302.rst

```diff
@@ -0,0 +1,67 @@
+::
+
+  ZIP: 302
+  Title: Standardized Memo Field Format
+  Owners: Jay Graber <jay@z.cash>
+  Status: Draft
+  Category: Standards
+  Created: 2017-02-08
+  License: MIT
+
+
+Abstract
+========
+
+This ZIP describes a proposed specification for a standardized format for displaying content within the encrypted memo field of shielded transactions.
+
+Motivation
+==========
+
+A well-defined standard for formatting content within the encrypted memo field would help expand the use cases by providing a structure for different types of data. Users and third-party services could benefit from a standardized formatting convention that defines the type and length of the data contained within.
+
+A proposed common use case could be a standard encoding for sender’s preferred return address.
+
+Specification
+===============
+
+This is a proposal to include the following memo field formatting conventions in the Zcash protocol specification:
+
++ If the first byte (byte 0)'s value is 0xF4 or smaller, then the reader should:
+     + strip any trailing zero bytes
+     + decode it as UTF-8 (replacing any incorrect UTF-8-encoded byte sequences with the replacement character U+FFFD), and display it to the user as a human-readable string.
++ If byte 0's value is 0xF5, then:
+     + Interpret the next few bytes (1 to 9 of them) as a 64-bit unsigned varint/ULEB, and use it as an arbitrary application-defined "type" field.
+     + Interpret the next bytes (1 to 2 of them) as a 16-bit unsigned ULEB, and use it as the length field.
```

Comment:
We should just use a 2-byte little-endian unsigned integer for the length field. There is no benefit to the variable-length encoding here, because the 1-byte length field will only occur for memo data at most 127 bytes, and the 2-byte encoding (either fixed or variable) will be necessary for any memo length near the 512-byte maximum where space optimisations might matter.

This draft ZIP has been up for a while, so there's a possible argument that someone may have already implemented this using the ULEB length field. I am unaware of any such implementation (`zcashd` certainly does not do so), and any associated use case should have been written up as a ZIP so that there was a corresponding allocated type. So I don't believe this is a problematic change, but I would be happy to see evidence to the contrary (e.g. link to an implementation).

Reply:
Hmm, actually I see that `zmsg` implements this with varints for both. Oh well, I guess it's fine.

Reply:
OK, we're going to leave the text as-is then.